enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,317 @@
+Since `UpstreamResolvers` doesn't allow for configuration of a custom zone, 
+it's assumed that the same TLS configuration will apply to _n_ `Upstream` 
+objects. 
+
+```go
+type UpstreamResolvers struct {
+  // ...<snip>
+
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). Populating a configmap name in this field
+  // effectively enables DNS-over-TLS. If this is empty or the configmap
+  // referenced does not exist, DNS-over-TLS will not be enabled.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or certificate bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  // 4. The ServerName value must be supplied when CABundle is configured.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
+
+  // serverName is the upstream server to connect to for DNS resolution.
+  // This is required when a CABundle is supplied. Populating the
+  // ServerName without a CABundle configured will have no effect.
+  //
+  // +optional
+  // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z0-9-\p{L}]){2,63})$`
+  ServerName string `json:"serverName,omitempty"`
+}
+```
+
+`Upstream` will get one new optional member:
+
+`EnableTLS` which allows an individual upstream to be opted into using
+DNS-over-TLS. This will be useful in cases where not all upstreams support
+DNS-over-TLS.
+
+```go
+type Upstream struct {
+  // ...snip
+
+  // enableTls allows cluster administrators to opt-in to using the TLS config in
+  // spec.upstreamResolvers for upstreams that support DNS-over-TLS.
+  //
+  // +optional
+  // +kubebuilder:validation:Enum=true;false
+  // +kubebuiler:default=false
+  EnableTLS string `json:"enableTls,omitempty"`
+}
+```
+
+The second API change will be to `ForwardPlugin`. `ForwardPlugin` will get 
+two new optional members:
+
+```go
+type ForwardPlugin struct {
+  // ...<snip>
+
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). Populating a configmap name in this field
+  // effectively enables DNS-over-TLS. If this is empty or the configmap
+  // referenced does not exist, DNS-over-TLS will not be enabled.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  // 4. The ServerName value must be supplied.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
```

Comment:
Same comment about granularity here.  Looks like the upstreams here may be a []string, so me have to create a new upstreams to represent this properly.

Reply:
Similarly here as with `UpstreamResolvers` in that we're limited by what we can represent in the Corefile. [These docs](https://docs.openshift.com/container-platform/4.9/networking/dns-operator.html#nw-dns-forward_dns-operator) have an example of the resulting Corefile when forwardplugin is configured but the main "gotcha" is that CoreDNS only allows one forward block per server so we can't configure separate TLS settings per upstream, they apply per server block (e.g. per set of zones that need forwarding).